File: engine/cli/audio_playback.py
"""Sequential audio playback for the CLI harnesses' TTS output.

Plays queued PCM16 chunks back-to-back on a background thread so translated
sentences never overlap or clip, and so playback never blocks the mic-capture
/ frame-processing loop.
"""

from __future__ import annotations

import queue
import threading

import numpy as np
import sounddevice as sd


class AudioPlayer:
    def __init__(self, device: int | None = None) -> None:
        self._device = device
        self._queue: queue.Queue[tuple[np.ndarray, int] | None] = queue.Queue()
        self._thread = threading.Thread(target=self._worker, daemon=True)
        self._thread.start()

    def enqueue(self, pcm16: bytes, sample_rate: int) -> None:
        if not pcm16:
            return
        audio = np.frombuffer(pcm16, dtype=np.int16)
        self._queue.put((audio, sample_rate))

    def _worker(self) -> None:
        while True:
            item = self._queue.get()
            if item is None:
                break
            audio, sample_rate = item
            sd.play(audio, samplerate=sample_rate, device=self._device, blocking=True)

    def close(self) -> None:
        self._queue.put(None)
        self._thread.join(timeout=2)

File: android/app/src/main/kotlin/com/interpreter/app/audio/AudioPlayback.kt
package com.interpreter.app.audio

import android.media.AudioAttributes
import android.media.AudioFormat
import android.media.AudioTrack
import java.util.concurrent.LinkedBlockingQueue
import java.util.concurrent.TimeUnit
import kotlin.concurrent.thread

/**
 * Plays queued translated-sentence audio back-to-back on a dedicated thread,
 * so sentences never overlap or clip and playback never blocks mic capture.
 */
class AudioPlayback {
    private val queue = LinkedBlockingQueue<Pair<ByteArray, Int>>()
    private var playbackThread: Thread? = null

    @Volatile
    private var isRunning = false

    fun start() {
        if (isRunning) return
        isRunning = true
        playbackThread = thread(name = "AudioPlaybackThread") {
            while (isRunning) {
                val item = queue.poll(200, TimeUnit.MILLISECONDS) ?: continue
                runCatching { playBlocking(item.first, item.second) }
            }
        }
    }

    fun enqueue(pcm16: ByteArray, sampleRate: Int) {
        if (pcm16.isEmpty()) return
        queue.put(pcm16 to sampleRate)
    }

    private fun playBlocking(pcm16: ByteArray, sampleRate: Int) {
        val minBufferSize = AudioTrack.getMinBufferSize(
            sampleRate,
            AudioFormat.CHANNEL_OUT_MONO,
            AudioFormat.ENCODING_PCM_16BIT,
        )
        val track = AudioTrack.Builder()
            .setAudioAttributes(
                AudioAttributes.Builder()
                    .setUsage(AudioAttributes.USAGE_VOICE_COMMUNICATION)
                    .setContentType(AudioAttributes.CONTENT_TYPE_SPEECH)
                    .build(),
            )
            .setAudioFormat(
                AudioFormat.Builder()
                    .setSampleRate(sampleRate)
                    .setEncoding(AudioFormat.ENCODING_PCM_16BIT)
                    .setChannelMask(AudioFormat.CHANNEL_OUT_MONO)
                    .build(),
            )
            .setBufferSizeInBytes(maxOf(minBufferSize, pcm16.size))
            .setTransferMode(AudioTrack.MODE_STATIC)
            .build()
        try {
            track.write(pcm16, 0, pcm16.size)
            track.play()
            val durationMs = (pcm16.size / 2) * 1000L / sampleRate // 2 bytes/sample
            Thread.sleep(durationMs)
        } finally {
            track.stop()
            track.release()
        }
    }

    fun stop() {
        isRunning = false
        playbackThread?.join(500)
        playbackThread = null
        queue.clear()
    }
}
